frontends/ast/simplify.cc

```diff
@@ -899,6 +899,73 @@ bool AstNode::simplify(bool const_fold, bool at_zero, bool in_lvalue, int stage,
 				if (!is_rand_reg)
 					log_warning("reg '%s' is assigned in a continuous assignment at %s:%d.%d-%d.%d.\n", children[0]->str.c_str(), filename.c_str(), location.first_line, location.first_column, location.last_line, location.last_column);
 			}
+
+			// Copy unpacked array
+			if (children[0]->id2ast->type == AST_MEMORY &&
+					children[1]->type == AST_IDENTIFIER && children[1]->id2ast && children[1]->id2ast->type == AST_MEMORY &&
+					children[0]->children.size() == 0 && children[1]->children.size() == 0) {
+				const auto& lhs_unpacked_range = children[0]->id2ast->children[1];
+				log_assert(lhs_unpacked_range->basic_prep);
+				const auto& rhs_unpacked_range = children[1]->id2ast->children[1];
+				log_assert(rhs_unpacked_range->basic_prep);
+
+				const auto lhs_width = lhs_unpacked_range->range_left - lhs_unpacked_range->range_right + 1;
+				const auto rhs_width = rhs_unpacked_range->range_left - rhs_unpacked_range->range_right + 1;
+
+				if (lhs_width != rhs_width)
+					log_error("Array '%s' and '%s' differs in size: %s:%d.%d-%d.%d.\n",
+						children[0]->str.c_str(), children[1]->str.c_str(), filename.c_str(),
+						location.first_line, location.first_column, location.last_line, location.last_column);
+
+				// Create template from assignment
+				const auto* tmpl = this->clone();
+
+				// Replace assignment with block
+				auto* block = this;
+				switch (type)
+				{
+				case AST_ASSIGN_EQ:
+				case AST_ASSIGN_LE:
+					block->type = AST_BLOCK;
+					break;
+
+				case AST_ASSIGN:
+					block->type = AST_GENBLOCK;
+					break;
+
+				default: // should not happen
+					log_assert(false);
+					break;
+				}
+				block->children.clear();
+
+				// Generate series of assignments
+				for (int i = 0 ; i < lhs_width ; ++i) {
+					auto* assign = tmpl->clone();
+					assign->children[0]->children.push_back(new AstNode(AST_RANGE));
+					assign->children[0]->children[0]->children.push_back(AstNode::mkconst_int(i + lhs_unpacked_range->range_right, true));
+
+					assign->children[1]->children.push_back(new AstNode(AST_RANGE));
+					assign->children[1]->children[0]->children.push_back(AstNode::mkconst_int(i + rhs_unpacked_range->range_right, true));
+
+					// simplify
+					assign->children[0]->basic_prep = false;
+					while (!assign->children[0]->basic_prep && assign->children[0]->simplify(false, false, true, stage, -1, false, in_param) == true)
```

Comment:
Is `== true` really needed here?

Reply:
It's not really needed, just followed the convention used in this file e.g:
```
while (!children[0]->basic_prep && children[0]->simplify(false, false, true, stage, -1, false, in_param) == true)
```
in line 873.

We can change this if you want, but then it'd good to update all the other `== true` lines so the code style is consistent in the file.

Reply:
There are only a handful of instances of `== true` in this file. My preference would be to avoid replicating that style in new code given many find it problematic. We may consider removing the other usages, but some may prefer avoiding the noise in the commit history.